Add G3 perf WA for Qwen2VL

nngokhale opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

nngokhale commented on optimum/habana/transformers/models/qwen2_vl/modeling_qwen2_vl.py:
```diff
@@ -589,9 +589,14 @@ def forward(
-                image_mask = (input_ids == self.config.image_token_id).unsqueeze(-1).expand_as(inputs_embeds)
                 image_embeds = image_embeds.to(inputs_embeds.device, inputs_embeds.dtype)
-                inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_embeds)
+                # G3 WA (masked_scatter has perf issue, flatten for hpu graphs)
```
Hi @imangohari1 
1. There is no loss in accuracy. This is just alternate way of putting image embeds into the input embeds.
2. It works better than masked_scatter on both G2 and G3. E.g., unit test with 5000 token embeds masked scatter on G2 takes 400ms, G3 it takes 6000ms whereas this WA takes 100ms on both G2/G3.

imangohari1 replied:
This block is completely rewritten in HF upstream starting 4.46.0:  https://github.com/huggingface/transformers/blob/v4.46.0/src/transformers/models/qwen2_vl/modeling_qwen2_vl.py#L1683C1-L1699C87

We are in process of updating to 4.49... so we should try to evaluate the HF upstream vs our changes. 

Can you please run this with the following changes and compare the results? 
Thanks. 

```suggestion
            if pixel_values is not None:
                pixel_values = pixel_values.type(self.visual.get_dtype())
                image_embeds = self.visual(pixel_values, grid_thw=image_grid_thw)
                n_image_tokens = (input_ids == self.config.image_token_id).sum().item()
                n_image_features = image_embeds.shape[0]
                if n_image_tokens != n_image_features:
                    raise ValueError(
                        f"Image features and image tokens do not match: tokens: {n_image_tokens}, features {n_image_features}"
                    )
                image_mask = (
                    (input_ids == self.config.image_token_id)
                    .unsqueeze(-1)
                    .expand_as(inputs_embeds)
                    .to(inputs_embeds.device)
                )
                image_embeds = image_embeds.to(inputs_embeds.device, inputs_embeds.dtype)
                inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_embeds)
```

imangohari1 reviewed (commented):
I think we need to try this with updated upstream code. see my comment below.

nngokhale commented:
> I think we need to try this with updated upstream code. see my comment below.

The upstream has changed only in a check from https://github.com/huggingface/transformers/commit/0f49deacbff3e57cde45222842c0db6375e4fa43
                 
                 n_image_tokens = (input_ids == self.config.image_token_id).sum().item()
                 n_image_features = image_embeds.shape[0]
                 if n_image_tokens != n_image_features:
                     raise ValueError(
                         f"Image features and image tokens do not match: tokens: {n_image_tokens}, features {n_image_features}"
                     )

The above check will not impact my change.

imangohari1 commented:
> > I think we need to try this with updated upstream code. see my comment below.
> 
> The upstream has changed only in a check from [huggingface/transformers@0f49dea](https://github.com/huggingface/transformers/commit/0f49deacbff3e57cde45222842c0db6375e4fa43)
> 
> ```
>              n_image_tokens = (input_ids == self.config.image_token_id).sum().item()
>              n_image_features = image_embeds.shape[0]
>              if n_image_tokens != n_image_features:
>                  raise ValueError(
>                      f"Image features and image tokens do not match: tokens: {n_image_tokens}, features {n_image_features}"
>                  )
> ```
> 
> The above check will not impact my change.

yes but I am wondering how that code as performs vs current pr changes. 
Please collect the following data and let's take a look at them:
- OH main as is, 
- OH main with PR 1884
- OH main with changes in upstream HF (excluding the PR changes). 

Thanks.

nngokhale commented:
> > > I think we need to try this with updated upstream code. see my comment below.
> > 
> > 
> > The upstream has changed only in a check from [huggingface/transformers@0f49dea](https://github.com/huggingface/transformers/commit/0f49deacbff3e57cde45222842c0db6375e4fa43)
> > ```
> >              n_image_tokens = (input_ids == self.config.image_token_id).sum().item()
> >              n_image_features = image_embeds.shape[0]
> >              if n_image_tokens != n_image_features:
> >                  raise ValueError(
> >                      f"Image features and image tokens do not match: tokens: {n_image_tokens}, features {n_image_features}"
> >                  )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The above check will not impact my change.
> 
> yes but I am wondering how that code as performs vs current pr changes. Please collect the following data and let's take a look at them:
> 
> * OH main as is,
> * OH main with PR 1884
> * OH main with changes in upstream HF (excluding the PR changes).
> 
> Thanks.

Tested with the following cmdline on G3 1.20:
python3 examples/image-to-text/run_pipeline.py --model_name_or_path Qwen/Qwen2-VL-2B-Instruct --image_path "https://llava-vl.github.io/static/images/view.jpg"  --use_hpu_graphs --bf16 --use_flash_attention 

Results:
OH main as is,
Output: The image shows a wooden dock extending into a calm lake. In the background, there are lush green forests and mountains, creating a serene and picturesque landscape. The sky is overcast, adding to the tranquil atmosphere.
04/01/2025 19:29:53 - INFO - __main__ - time = 1345.3735169954598ms, Throughput (including tokenization) = 23.04192821427792 tokens/second

OH main with PR 1884
Output: The image shows a wooden dock extending into a calm lake. In the background, there are trees and mountains, creating a serene and picturesque landscape. The sky is overcast, adding to the tranquil atmosphere.
04/01/2025 19:36:52 - INFO - __main__ - time = 478.12585642095655ms, Throughput (including tokenization) = 64.83648517160857 tokens/second

OH main with changes in upstream HF (excluding the PR changes). see https://github.com/nngokhale/optimum-habana/commits/mainup/
Output: The image shows a wooden dock extending into a calm lake. In the background, there are trees and mountains, creating a serene and picturesque landscape. The sky is overcast, adding to the tranquil atmosphere.
04/01/2025 19:33:46 - INFO - __main__ - time = 1289.4892164040357ms, Throughput (including tokenization) = 25.591528475147875 tokens/second

imangohari1 reviewed (changes requested):
Hi @nngokhale 
I did some testing on both g2/g3 with and without this PR, with and without hpu-graphs. 
Results are below. 
I think perf is good. 
Please apply the patch I attached to this comment with `git am < 0001*` and push the commit before a final review. 
[0001-fea-make-style.-Added-original-code-ref.patch](https://github.com/user-attachments/files/19572238/0001-fea-make-style.-Added-original-code-ref.patch)

thanks. 

## Test 1 (no hpu graph) 
cmd:
```
python examples/image-to-text/run_pipeline.py --model_name_or_path Qwen/Qwen2-VL-7B-Instruct  --bf16 --sdp_on_bf16
```
G2
oh main@1e0aa86a58884c46b7b8448b8bbf9654e3d816eb
```
04/02/2025 15:45:41 - INFO - __main__ - time = 5125.258138595382ms, Throughput (including tokenization) = 12.292063793935197 tokens/second
```
pr 1884@2c8113772d48ccc67de21dd3a6c78c7996a65855
```
04/02/2025 15:56:23 - INFO - __main__ - time = 4316.859199799364ms, Throughput (including tokenization) = 12.509094575637253 tokens/second
```

G3
oh main@1e0aa86a58884c46b7b8448b8bbf9654e3d816eb
```
04/02/2025 16:14:03 - INFO - __main__ - time = 4063.493970176205ms, Throughput (including tokenization) = 13.289056264468481 tokens/second
```
pr 1884@2c8113772d48ccc67de21dd3a6c78c7996a65855
```
04/02/2025 16:18:39 - INFO - __main__ - time = 3197.8034294210374ms, Throughput (including tokenization) = 16.886591434351143 tokens/second
```
## Test 2 (with hpu graphs) 
cmd: 
```
python examples/image-to-text/run_pipeline.py --model_name_or_path Qwen/Qwen2-VL-7B-Instruct  --image_path "https://llava-vl.github.io/static/images/view.jpg" --use_hpu_graphs --bf16 --use_flash_attention
```

G2
oh main@1e0aa86a58884c46b7b8448b8bbf9654e3d816eb
```
04/02/2025 16:04:47 - INFO - __main__ - time = 644.4136726029683ms, Throughput (including tokenization) = 48.105745296778494 tokens/second
```
pr 1884@2c8113772d48ccc67de21dd3a6c78c7996a65855
```
04/02/2025 16:03:24 - INFO - __main__ - time = 578.834522201214ms, Throughput (including tokenization) = 53.55589345658241 tokens/second
```

G3:
oh main@1e0aa86a58884c46b7b8448b8bbf9654e3d816eb
```
04/02/2025 16:15:27 - INFO - __main__ - time = 1335.7813362032175ms, Throughput (including tokenization) = 23.207391179841917 tokens/second
```
pr 1884@2c8113772d48ccc67de21dd3a6c78c7996a65855
```
04/02/2025 16:19:52 - INFO - __main__ - time = 460.3163016028702ms, Throughput (including tokenization) = 67.3449971075426 tokens/second
```

nngokhale commented:
> Hi @nngokhale I did some testing on both g2/g3 with and without this PR, with and without hpu-graphs. Results are below. I think perf is good. Please apply the patch I attached to this comment with `git am < 0001*` and push the commit before a final review. [0001-fea-make-style.-Added-original-code-ref.patch](https://github.com/user-attachments/files/19572238/0001-fea-make-style.-Added-original-code-ref.patch)

Thanks, Done.

imangohari1 reviewed (approved):
LGTM! @regisss  please review.

regisss commented:
@nngokhale Can you see if this applicable to the `transformers_future` branch (upgrade to Transformers 4.49) that is going to be merged today or tomorrow?

nngokhale commented:
> 4.49

Yes it is still applicable. We tested with relevant latest model upstream code and found that change is still needed for performance.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_1884). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

regisss reviewed (approved):
LGTM!
